src/vs/workbench/contrib/textEditor/browser/textEditorService.ts

```diff
@@ -0,0 +1,246 @@
+import { EditorInput } from 'vs/workbench/common/editor/editorInput';
+import { INewUntitledTextEditorOptions, IUntitledTextEditorService } from 'vs/workbench/services/untitled/common/untitledTextEditorService';
+import { Schemas } from 'vs/base/common/network';
+import { DiffEditorInput } from 'vs/workbench/common/editor/diffEditorInput';
+import { SideBySideEditorInput } from 'vs/workbench/common/editor/sideBySideEditorInput';
+import { TextResourceEditorInput } from 'vs/workbench/common/editor/textResourceEditorInput';
+import { UntitledTextEditorInput } from 'vs/workbench/services/untitled/common/untitledTextEditorInput';
+import { IUntitledTextEditorModel } from 'vs/workbench/services/untitled/common/untitledTextEditorModel';
+import { basename } from 'vs/base/common/resources';
+import { URI } from 'vs/base/common/uri';
+import { IUriIdentityService } from 'vs/platform/uriIdentity/common/uriIdentity';
+import { IFileService } from 'vs/platform/files/common/files';
+import { IEditorResolverService, RegisteredEditorPriority } from 'vs/workbench/services/editor/common/editorResolverService';
+import { Disposable } from 'vs/base/common/lifecycle';
+import { ITextEditorService } from 'vs/workbench/services/textfile/common/textEditorService';
+import { MergeEditorInput } from 'vs/workbench/contrib/mergeEditor/browser/mergeEditorInput';
+
+export class TextEditorService extends Disposable implements ITextEditorService {
+
+	declare readonly _serviceBrand: undefined;
+
+	private readonly editorInputCache = new ResourceMap<TextResourceEditorInput | IFileEditorInput | UntitledTextEditorInput>();
+
+	private readonly fileEditorFactory = Registry.as<IEditorFactoryRegistry>(EditorExtensions.EditorFactory).getFileEditorFactory();
+
+	constructor(
+		@IUntitledTextEditorService private readonly untitledTextEditorService: IUntitledTextEditorService,
+		@IInstantiationService private readonly instantiationService: IInstantiationService,
+		@IUriIdentityService private readonly uriIdentityService: IUriIdentityService,
+		@IFileService private readonly fileService: IFileService,
+		@IEditorResolverService private readonly editorResolverService: IEditorResolverService
+	) {
+		super();
+
+		this.registerTextEditors();
+	}
+
+	private registerTextEditors(): void {
+		this._register(this.editorResolverService.registerEditor(
+			'*',
+			{
+				id: DEFAULT_EDITOR_ASSOCIATION.id,
+				label: DEFAULT_EDITOR_ASSOCIATION.displayName,
+				detail: DEFAULT_EDITOR_ASSOCIATION.providerDisplayName,
+				priority: RegisteredEditorPriority.builtin
+			},
+			{},
+			{
+				createEditorInput: editor => ({ editor: this.createTextEditor(editor) }),
+				createUntitledEditorInput: untitledEditor => ({ editor: this.createTextEditor(untitledEditor) }),
+				createDiffEditorInput: diffEditor => ({ editor: this.createTextEditor(diffEditor) }),
+				createMergeEditorInput: mergeEditor => ({ editor: this.createTextEditor(mergeEditor) })
```

Comment:
@jrieken yes, I had a follow up discussion with Logan where going with the sparse solution was not possible. But I let him comment on that.

Reply:
I do have a PR out for switching to objects as discussed. After some exploration and discussion with Ben it appeared that supporting separate contributions within the resolver would produce quite a bit of technical debt which we worked so hard to cleanup when making the resolver a clean and straightforward model. It also sounded like the main concern during the meeting was not how the resolver registers things but which layer these things belong in. In the future I think the plan would be to move all editor registrations to contrib so we figured if it simplifies this use case a bunch what's a better exercise than starting now. Technically if we wanted to the registration object could be exposed in a central place and then `createMergeEditorInput` property could be set somewhere else, but I really like how simple and straightforward this is. Diff, and Merge are not separate editor types they're just classes of text / notebook editors so keeping it with the text, notebook, custom etc registration seems to fit nicely in what I would define as editor classes (diff, single resource, untitlted) vs editor types (text, custom, notebook).

Reply:
> appeared that supporting separate contributions within the resolver would produce quite a bit

I don't believe that contributions within a single resolver was the ask. Instead the desire is to have a separate resolver that only handles merge editors and that isn't aware of diff, untitled, or other current and future editor classes. This is about "separation of concerns" and not necessarily about the folder in which the code is stored. 

I want to have (inside the merge-contrib) something like this

```ts
this.editorResolverService.registerEditor(
 '*',
  {
 	id: 'textmerge',
 	label: 'Text Merge Editor',
 	detail: '',
 	priority: RegisteredEditorPriority.builtin
 },
 {},
 {
 	createEditorInput: undefined,
 	createUntitledEditorInput: undefined,
 	createDiffEditorInput: undefined,
 	createMergeEditorInput: mergeEditor => { /*create 3wm editor*/ }
}
```

My expectation is that for merge-inputs my resolver is taken into consideration but that it can also be over-shadowed by another merge-handler of another resolver.

Reply:
But it's still just a different form of the text editor so it should be a part of the text editor resolution if possible. We also do not allow `createEditorInput` to be undefined because it is always the last factory called if nothing else fits. This is to ensure that something is always constructable. The merge editor is not a custom type like a notebook or webview but is  a whole new resource representation which each editor type should have the option to specify a factory for